x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,244 @@
+
+    private byte[] toByteArray(InputStream is) throws IOException {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        byte[] internalBuffer = new byte[1024];
+        int read = is.read(internalBuffer);
+        while (read != -1) {
+            baos.write(internalBuffer, 0, read);
+            read = is.read(internalBuffer);
+        }
+        return baos.toByteArray();
+    }
+
+    protected URL createAPIKeyUrl() throws MalformedURLException, URISyntaxException {
+        return new URL(defaultUrl, (defaultUrl.toURI().getPath() + "/_security/api_key").replaceAll("/+", "/"));
+    }
+
+    protected URL getHttpInfoUrl() throws MalformedURLException, URISyntaxException {
+        return new URL(defaultUrl, (defaultUrl.toURI().getPath() + "/_nodes/http").replaceAll("/+", "/"));
+    }
+
+    protected static List<String> getBoundAddresses(Map<?, ?> nodesInfo) {
+        nodesInfo = (Map<?, ?>) nodesInfo.get("nodes");
+        Map<?, ?> nodeInfo = (Map<?, ?>) nodesInfo.values().iterator().next();
+        Map<?, ?> http = (Map<?, ?>) nodeInfo.get("http");
+        return (ArrayList<String>) http.get("bound_address");
+    }
+
+    static String getVersion(Map<?, ?> nodesInfo) {
+        nodesInfo = (Map<?, ?>) nodesInfo.get("nodes");
+        Map<?, ?> nodeInfo = (Map<?, ?>) nodesInfo.values().iterator().next();
+        return nodeInfo.get("version").toString();
+    }
+
+    protected String getApiKey(String user, SecureString password) throws Exception {
+        final CheckedSupplier<String, Exception> createApiKeyRequestBodySupplier = () -> {
+            XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+            xContentBuilder.startObject()
+                .field("name", "enrollment_token_API_key_" + UUIDs.base64UUID())
+                .field("expiration", ENROLL_API_KEY_EXPIRATION)
+                .startObject("role_descriptors")
+                .startObject("create_enrollment_token")
+                .field("cluster", "[cluster:admin/xpack/security/enroll/node]")
+                .endObject()
+                .endObject()
+                .endObject();
+            return Strings.toString(xContentBuilder);
+        };
+
+        final URL Url = createAPIKeyUrl();
+        final HttpResponse httpResponseApiKey = client.execute("POST", Url, user, password,
+            createApiKeyRequestBodySupplier, is -> responseBuilder(is));
+        final int httpCode = httpResponseApiKey.getHttpStatus();
+
+        if (httpCode != HttpURLConnection.HTTP_OK) {
+            logger.error("Error " + httpCode + "when calling GET " + Url + ". ResponseBody: " +
+                (httpResponseApiKey == null ? "" : httpResponseApiKey.getResponseBody()));
+            throw new IllegalStateException("Unexpected response code [" + httpCode + "] from calling POST "
+                + Url);
+        }
+
+        final String apiKey = Objects.toString(httpResponseApiKey.getResponseBody().get("api_key"), "");
```

Comment:
Because of reasons you must also look into the `"id"` field to assemble the credential: https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-api-key.html .